Update Web OTP for origin trial graduation (#2442)
GoogleChrome/web.dev: src/site/content/en/blog/web-otp/index.md

commonly using phone numbers as an identifier for users of their services.

There are a variety of ways to verify phone numbers, but a randomly generated
one-time password (OTP) sent by SMS is one of the most common. Sending this code
back to the developer's server demonstrates control of the phone number.

{% Aside %}
The Web OTP API was originally called the SMS Receiver API. You may still see it

> named that way in some places. If used that API, you should still read this

Reviewer: ```suggestion
named that way in some places. If you used that API, you should still read this
```

1. Create explainer
</td>
<td markdown="block">

> <a href="https://github.com/samuelgoto/sms-receiver/blob/master/README.md">Complete</a>

Reviewer 2: This should be `https://github.com/WICG/WebOTP/blob/master/explainer.md`

<strong>5. Launch</strong>
</td>
<td markdown="block">
<strong>Chrome 83</strong>
</td>
</tr>
</table>

> ## Changes from earlier versions

Author: @agektmr FYI

Reviewer 2: LGTM

sends a text message to the user over SMS and the user enters the OTP from the
message to verify the ownership of the phone number.

With the Web OTP API, these steps are as easy as one tap for the user, as
demonstrated in the video. When the text message arrives, a bottom sheet pops up
and prompts the user to verify their phone number. After clicking the **Verify**
button on the bottom sheet, the browser pastes the OTP into the form and the
form is submitted without the user needing to press **Continue**.

<video autoplay loop muted playsinline>

>   <source src="https://storage.googleapis.com/web-dev-assets/sms-receiver-announce/demo.mp4" type="video/mp4">

Reviewer 2: Let's replace the video. I can work on this.

Author: Thanks you.

Reviewer 2: The video is replaced.

>   <figcaption class="w-figcaption w-figcaption--fullbleed">

Reviewer 2: We should replace the diagram image as well. I'll work on it.

Author: Thank you.

Reviewer 2: The image is replaced.

That is how the Web OTP API works for users.

{% Aside 'caution' %}

> If you are using a work profile on your Android device and the Web OTP does
> not seem to be working, try installing and using Chrome on your personal profile

Author: ```suggestion
If you are using a work profile on your Android device and the Web OTP does
not work, try installing and using Chrome on your personal profile
```

  signal.abort();
}, 10 * 1000);

const content = await navigator.credentials.get({
  otp: { transport:['sms'] },
  abort: signal
});
```

### Use the API declaratively

The code below demonstrates a web component that extends `input`.

```js
customElements.define("one-time-code",
  class extends HTMLInputElement {
    connectedCallback() {
      this.receive();
    }
    async receive() {
      try {
        const sms = await navigator.credentials.get({
          otp: { transport:['sms'] }
        });
        this.value = sms.code;
        this.form.submit();
      } catch (e) {
        console.error(e);
      }
    }
  }, {
    extends: "input"
});
```
After this declaration you can add `is="one-time-code` to any `input` element.
As soon as the element is added to the document tree, it waits for an SMS
message to arrive and submits the form as soon as the OTP is passed.

```html
<form>
  <input is="one-time-code" autocomplete="one-time-code" required />
  <input type="submit">
</form>
```

### Format the SMS message {: #format }

The API itself should look simple enough, but a critical part is to
format your SMS text message according to a specific convention. The message has

> to be sent after `avigator.credentials.get()` is called and must comply with the

Reviewer 2: `navigator.credentials.get()`

formatting convention.

> The SMS message must be received on the device where `avigator.credentials.get()`

Reviewer 2: `navigator.credentials.get()`

was called.

The message must adhere to the following formatting:

* The origin part of the URL of the website that invoked the API must be
  preceded by `@`.
* The URL must contain (for [the time
  being](https://github.com/samuelgoto/sms-receiver/issues/4#issuecomment-528991114))
  a query parameter whose value is the application hash of the user's Chrome
  instance. See [Use the application hash](#use-the-hash) for a list of valid hashes.
* The URL must contain a pound sign ('`#`') followed by the OTP.
* Optionally, the message may contain additional text for the user.

For example:

```text
Your OTP is: 123456.

@https://www.example.com #123456&xFJnfg75+8v
```

> ### Use the application hash {: #use-the-hash }

Reviewer 2: We can now get rid of app hash section. It's no longer needed.

Author: I almost did, but it's needed until at least when 83 goes to stable. Does the flag in 83 default to on or off?

Reviewer 2: The flag will not be required on 83.

Reviewer 2: I edited it.

Feature detection is the same as for many other APIs:

```js
if ('OTPCredential' in window) {
  ...
}
```

### Process the OTP

The Web OTP API itself is simple enough. Use

> [`navigator.credentials.get()`](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/get)

Reviewer: ```suggestion
[`navigator.credentials.get()`](https://developer.mozilla.org/docs/Web/API/CredentialsContainer/get)
```

to obtain the OTP. Web OTP adds a new `otp` option to that method. It only has
one property: `transport`, whose value must be an array with the string `'sms'`.

```js
const content = await navigator.credentials.get({
  otp: {
    transport:['sms']
  }
});
```

This triggers the browser's permission flow when an SMS message arrives. If permission is
granted, the returned promise resolves with an `OTPCredential` object.

```json
{
  code: "123456" // Obtained OTP
  type: "otp"  // `type` is always "otp"
}
```

Next, pass the OTP value to an `input` field and submit it on behalf of the user.

```js
document.querySelector('#input').value = content.code;
```

### Aborting the message {: #aborting }

To set a timeout that aborts the `get()` call, pass an `AbortController`
instance in the [`options`

> object](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/get#Parameters).

Reviewer: ```suggestion
object](https://developer.mozilla.org/docs/Web/API/CredentialsContainer/get#Parameters).
```

```js
const signal = new AbortController();
let timer = setTimeout(() => {
  signal.abort();
}, 10 * 1000);

const content = await navigator.credentials.get({
  otp: { transport:['sms'] },
  abort: signal
});
```

### Use the API declaratively

> The code below demonstrates a web component that extends `input`.

Reviewer: ```suggestion
The code below demonstrates a Web Component that extends `input`.
```

```js
if ('customElements' in window && 'OTPCredential' in window) {
  customElements.define("one-time-code",
    class extends HTMLInputElement {
      connectedCallback() {
        this.signal = new AbortController();
        this.receive();
      }
      disconnectedCallback() {
        this.abort();
      }
      abort() {
        this.signal.abort();
      }
      async receive() {
        try {
          const content = await navigator.credentials.get({
            otp: {transport:['sms']}, abort: this.signal
          });
          this.value = content.code;
          this.dispatchEvent(new Event('autocomplete'));
        } catch (e) {
          console.error(e);
        }
      }
    }, {
      extends: "input"
  });
}
```

> After this declaration you can add `is="one-time-code` to any `input` element.

Reviewer: ```suggestion
After this declaration you can add `is="one-time-code"` to any `input` element.
```

As soon as the element is added to the document tree, it starts waiting
for SMS messages to arrive and emits an `autocomplete` event as soon
as an OTP is passed.

```html
<form>

>   <input is="one-time-code" autocomplete="one-time-code" required />

Reviewer: ```suggestion
  <input is="one-time-code" autocomplete="one-time-code" required/>
```

});
```

> This triggers the browser's permission flow as an SMS arrives. If permission is

Author: ```suggestion
This triggers the browser's permission flow when an SMS message arrives. If permission is
```

  });
}
```

After this declaration you can add `is="one-time-code` to any `input` element.

> As soon as the element is added to the document tree, it will start waiting
> for SMS to arrive and emit an `autocomplete` event as soon as an OTP is passed.

Author: ```suggestion
As soon as the element is added to the document tree, it starts waiting
for SMS messages to arrive and emits an `autocomplete` event as soon
as an OTP is passed.
```

* The `get()` receives only the OTP rather than the entire SMS message as
  `receive()` did before.
* It's now possible to [abort the call to `get()`](#aborting).
* It no longer requires an app hash string to be included in the SMS message.

## FAQ
### Is this API compatible between different browsers?

> Chromium and WebKit at least agreed on the SMS text message format. Find WebKit's

Author: ```suggestion
Chromium and WebKit agreed on the SMS text message format. Find WebKit's
```

Are you planning to use the Web OTP API? Your public support helps us prioritize
features, and shows other browser vendors how critical it is to support them.
Send a Tweet to [@ChromiumDev](https://twitter.com/chromiumdev) with
`#webotp` and let us know where and how you're using it.

## Differences from SMS Receiver API {: #differences }
Consider Web OTP API an evolved version of the SMS Receiver API. Web OTP API has
a few significant differences compared to the SMS Receiver API.

* The [expected text format](#format) for the SMS message has changed.
* It no longer requires an app hash string to be included in the SMS message.
* The method called is now `navigator.credentials.get()` rather than
  `navigator.sms.receive()`.
* The `get()` receives only the OTP rather than the entire SMS message as
  `receive()` did before.
* It's now possible to [abort the call to `get()`](#aborting).

## FAQ
### Is this API compatible between different browsers?

Chromium and WebKit agreed on the SMS text message format. Find WebKit's
documentation here:

> [https://github.com/WebKit/explainers/tree/master/sms-one-time-code-format](https://github.com/WebKit/explainers/tree/master/sms-one-time-code-format)

Reviewer: ```suggestion
[Delivering origin-bound one-time codes over SMS](https://github.com/WebKit/explainers/tree/master/sms-one-time-code-format)
```
